packages/react-dom/src/__tests__/refs-destruction-test.js

```diff
@@ -46,77 +54,96 @@ describe('refs-destruction', () => {
+    expect(theInnerDivRef.current).toBe(null);
+    expect(theInnerClassComponentRef.current).toBe(null);
   });
 
-  it('should remove refs when destroying the child', () => {
+  it('should remove refs when destroying the child', async () => {
     const container = document.createElement('div');
-    const testInstance = ReactDOM.render(<TestComponent />, container);
-    expect(
-      ReactTestUtils.isDOMComponent(testInstance.theInnerDivRef.current),
-    ).toBe(true);
-    expect(testInstance.theInnerClassComponentRef.current).toBeTruthy();
+    const root = ReactDOMClient.createRoot(container);
+    await act(async () => {
+      root.render(<TestComponent />);
+    });
+
+    expect(ReactTestUtils.isDOMComponent(theInnerDivRef.current)).toBe(true);
+    expect(theInnerClassComponentRef.current).toBeTruthy();
 
-    ReactDOM.render(<TestComponent destroy={true} />, container);
+    await act(async () => {
+      root.render(<TestComponent destroy={true} />);
+    });
 
-    expect(testInstance.theInnerDivRef.current).toBe(null);
-    expect(testInstance.theInnerClassComponentRef.current).toBe(null);
+    expect(theInnerDivRef.current).toBe(null);
+    expect(theInnerClassComponentRef.current).toBe(null);
   });
 
-  it('should remove refs when removing the child ref attribute', () => {
+  it('should remove refs when removing the child ref attribute', async () => {
     const container = document.createElement('div');
-    const testInstance = ReactDOM.render(<TestComponent />, container);
+    const root = ReactDOMClient.createRoot(container);
+    await act(async () => {
+      root.render(<TestComponent />);
+    });
 
-    expect(
-      ReactTestUtils.isDOMComponent(testInstance.theInnerDivRef.current),
-    ).toBe(true);
-    expect(testInstance.theInnerClassComponentRef.current).toBeTruthy();
+    expect(ReactTestUtils.isDOMComponent(theInnerDivRef.current)).toBe(true);
+    expect(theInnerClassComponentRef.current).toBeTruthy();
 
-    ReactDOM.render(<TestComponent removeRef={true} />, container);
+    await act(async () => {
+      root.render(<TestComponent removeRef={true} />);
+    });
 
-    expect(testInstance.theInnerDivRef.current).toBe(null);
-    expect(testInstance.theInnerClassComponentRef.current).toBe(null);
+    expect(theInnerDivRef.current).toBe(null);
+    expect(theInnerClassComponentRef.current).toBe(null);
   });
 
-  it('should not error when destroying child with ref asynchronously', () => {
+  it('should not error when destroying child with ref asynchronously', async () => {
+    let nestedRoot;
     class Modal extends React.Component {
       componentDidMount() {
         this.div = document.createElement('div');
+        nestedRoot = ReactDOMClient.createRoot(this.div);
         document.body.appendChild(this.div);
         this.componentDidUpdate();
       }
 
       componentDidUpdate() {
-        ReactDOM.render(<div>{this.props.children}</div>, this.div);
+        setTimeout(() => {
```

Comment:
This is clowning af but the test before was clowny af (rendering and unmounting roots in lifecycles) so I don't want to rewrite it and risk losing whatever it was testing for.

Reply:
Needs to be flushSync because you can't use async/await in lifecycles. Needs to be a timeout because we warn if you use flushSync in lifecycles. The act flush ensures the timeout is fired.

Reply:
I don't really understand what's being tested here. We could also mark it as legacy and gate the tests.

Reply:
It's testing that if you have a nested root that you animate on unmount, there's no error. This is how you did portals before createPortal, so I don't think it's legacy until we remove class components.

Reply:
@kassens but I'm curious what you mean by gating, if this was legacy what gate would we use?